Mathlib/NumberTheory/LSeries/Deriv.lean

```diff
@@ -0,0 +1,227 @@
+  the abscissa of absolute convergence of `f` (`LSeries.hasDerivAt`) and that its derivative
+  there is the negative of the `LSeries` of the point-wise product `log * f` (`LSeries.deriv`).
+
+* We prove similar results for iterated derivatives (`LSeries.iteratedDeriv`).
+
+* We use this to show that `LSeries f` is holomorphic on the right half-plane of
+  absolute convergence (`LSeries.analyticOn`).
+
+## Implementation notes
+
+We introduce `LSeries.logMul` as an abbreviation for the point-wise product `log * f`, to avoid
+the problem that this expression does not type-check. Similarly, we make use of the abbreviation
+`LSeries.logPowMul n f` for `log^n * f`.
+-/
+
+-- Should this go into its own file (because it needs `EReal`s, it doesn't fit well
+-- in some existing file), together with versions for the left, upper and lower half-planes?
+/-- An open right half-plane is open in the complex plane. -/
+lemma Complex.isOpen_rightHalfPlane (x : EReal) : IsOpen {z : ℂ | x < z.re} :=
+  isOpen_lt continuous_const <| EReal.continuous_coe_iff.mpr continuous_re
+
+open Complex LSeries
+
+/-!
+### Multiplication by (powers of) log
+-/
+
+/-- The (point-wise) product of `log : ℕ → ℂ` with `f`. -/
+noncomputable abbrev LSeries.logMul (f : ℕ → ℂ) (n : ℕ) : ℂ := log n * f n
+
+
+/-- The (point-wise) product of the `m`th power of `log` with `f`. -/
+noncomputable abbrev logPowMul (m : ℕ) (f : ℕ → ℂ) (n : ℕ) : ℂ :=
+  (log n) ^ m * f n
+
+lemma logPowMul_zero (f : ℕ → ℂ) : logPowMul 0 f = f := by
+  ext
+  simp only [logPowMul, pow_zero, one_mul]
+
+lemma logPowMul_succ (m : ℕ) (f : ℕ → ℂ) : logPowMul m.succ f = logMul (logPowMul m f) := by
+  ext
+  simp only [logPowMul, pow_succ, mul_assoc, logMul]
+
+lemma logPowMul_succ' (m : ℕ) (f : ℕ → ℂ) : logPowMul m.succ f = logPowMul m (logMul f) := by
+  ext m
+  simp only [logPowMul, pow_succ, logMul, ← mul_assoc]
+  rw [mul_comm (log m)]
+
+/-!
+### The derivative of an L-series
+-/
+
+/-- The derivative of the terms of an L-series. -/
+lemma LSeries.hasDerivAt_term (f : ℕ → ℂ) (n : ℕ) (s : ℂ) :
+    HasDerivAt (fun z ↦ term f z n) (-(term (logMul f) s n)) s := by
+  rcases n.eq_zero_or_pos with rfl | hn
+  · simp only [term_zero, neg_zero, hasDerivAt_const]
+  simp only [term_of_ne_zero hn.ne']
+  rw [← neg_div, ← neg_mul, mul_comm, mul_div_assoc]
+  simp_rw [div_eq_mul_inv, ← cpow_neg]
+  refine HasDerivAt.const_mul (f n) ?_
+  rw [mul_comm, ← mul_neg_one (Complex.log n), ← mul_assoc]
+  exact (hasDerivAt_neg' s).const_cpow (Or.inl <| Nat.cast_ne_zero.mpr hn.ne')
+
+/-- The derivative of the terms of an L-series. -/
+lemma LSeries.deriv_term (f : ℕ → ℂ) (n : ℕ) (s : ℂ) :
+    deriv (fun z ↦ term f z n) s = -(term (logMul f) s n) :=
+  (hasDerivAt_term f n s).deriv
+
+/-- The derivative of the terms of an L-series. -/
+lemma LSeries.deriv_term' (f : ℕ → ℂ) (n : ℕ) :
+    deriv (fun z ↦ term f z n) = fun s ↦ -(term (logMul f) s n) :=
+  funext <| deriv_term f n
```

Comment:
How useful is this, really? I'm a bit sceptical of the value of statements of the form `deriv f = foo`, since virtually any non-trivial place where you use them will need a differentiability hypothesis anyway, and if you prove that, you've just proved the two separate parts of `HasDerivAt f foo`.

Reply:
The first is used in the proof of `LSeries.LSeriesSummable_logMul_and_hasDerivAt`: one needs to know the derivatives of the terms in the sum. One could replace it by its proof there, I assume.